### ISSUE TITLE
2 tasks on many large graphs

### ISSUE THREAD

**Issue by jlevy44** [rusty1s/pytorch_geometric#275]
I'm not sure if this is possible, but I have 500 large graphs with about 20k nodes each, with no features (just set one column equal to ones). I split these graphs into 75/15/10 train/val/test. Each node in each graph has a class label, and each graph has a continuous value from 0-1. I want to accomplish 2 tasks.

1. Node level classification: I had been doing this previously on single graphs, but now I want to train off of entire graphs and predict on each node for unseen graphs.
2. Graph level regression: Now I'd like to use MLPs on graph level representations to predict a value between 0 and 1 for the entire graph. 

Right now, I've implemented a modified graphSAGE, after the concatenation step, for node level predictions, it passes the concatenated embedding through an mlp; for graph level predictions, it performs global mean poloing, then through an MLP with sigmoid output.

Am I using the right approach? What other considerations should I be making or should I just be using normal GCN layers instead of graphSAGE?

**Comment by rusty1s** [comment]
This is the right approach. However, I would use GIN layers + Jumping Knowledge because of their superior performance.

**Comment by jlevy44** [comment]
Okay. I'll give it a shot.

**Comment by jlevy44** [comment]
Where would I apply jumping knowledge? 

I think I'd use GIN in place of graphSAGE, is JK for the aggregation in place of global mean pool?

**Comment by rusty1s** [comment]
No before that as a layerwise aggregation.

**Comment by jlevy44** [comment]
Oh okay. Can you point me to an example?

Or some pseudocode illustrating the flow?

**Comment by jlevy44** [comment]
I think I see.

So GIN -> JK -> Global Mean Pool

**Comment by jlevy44** [comment]
I don't entirely see how jumping knowledge is different than what is implemented in GraphSAGE. Seems JK is just concatenating along an axis, graphsage there is also this concatenation during the forward pass.

**Comment by jlevy44** [comment]
I suppose if I used LSTM.

**Comment by jlevy44** [comment]
Actually, let me read more into your code.

**Comment by rusty1s** [comment]
Oh ok, I guess you already implemented it (check the benchmark.kernel nets). Sorry for the confusion.

**Comment by jlevy44** [comment]
Oh no, I just implemented graphsage.

Here's what I have so far for GIN (help from your code):
```
class GIN(torch.nn.Module):
	def __init__(self, dataset, num_layers, hidden):
		super(GIN, self).__init__()
		self.conv1 = GINConv(
			Sequential(
				Linear(dataset.num_features, hidden),
				ReLU(),
				Linear(hidden, hidden),
				ReLU(),
				BN(hidden),
			),
			train_eps=True)
		self.convs = torch.nn.ModuleList()
		for i in range(num_layers - 1):
			self.convs.append(
				GINConv(
					Sequential(
						Linear(hidden, hidden),
						ReLU(),
						Linear(hidden, hidden),
						ReLU(),
						BN(hidden),
					),
					train_eps=True))
		self.agg = JumpingKnowledge('cat')
		self.lin1 = torch.nn.Linear(num_layers * hidden, hidden)
		self.altlin = torch.nn.Linear(hidden, hidden)
		self.lin2 = Linear(hidden, dataset.y.size()[1] if len(dataset.y.size())>1 else 1)
		#self.lin2 = Linear(hidden, dataset.num_classes)

	def reset_parameters(self):
		self.conv1.reset_parameters()
		for conv in self.convs:
			conv.reset_parameters()
		self.lin1.reset_parameters()
		self.lin2.reset_parameters()

	def forward(self, x, edge_index, batch, graph_level=False):
		#x, edge_index, batch = data.x, data.edge_index, data.batch
		if edge_index.size()[0] >2:
			edge_index=torch.t(edge_index)
		x=x.float()
		x = self.conv1(x, edge_index)
		xs = [x]
		for conv in self.convs:
			x = conv(x, edge_index)
			xs += [x]
		z = self.agg(xs)
		if graph_level:
			z = global_mean_pool(z, batch)
			x = F.relu(self.lin1(z))
			x = F.dropout(x, p=0.5, training=self.training)
			x = self.lin2(x)
			return self.sigmoid(x), z # F.log_softmax(x, dim=-1)
		else:
			x=F.relu(self.altlin(x))
			return self.lin2(x), z

```

**Comment by jlevy44** [comment]
Is this correct for the tasks I have described?

**Comment by rusty1s** [comment]
Looks good to me.

**Comment by jlevy44** [comment]
Seems like nothing is really training though. I'm using BCELoss (MSE gives same results). The model stops training after the first epoch. Here's an example of my training curve. I've been wondering if this is due to my architecture versus my training set.

![training_curve](https://user-images.githubusercontent.com/19698023/57171706-e268c980-6de4-11e9-956e-0d438eac69b3.png)

**Comment by rusty1s** [comment]
You could try a different dataset or a simpler model.

**Comment by rusty1s** [comment]
I'm closing this, but feel free to discuss this here further anyway.
